### ISSUE TITLE
cpp compilation issue with ceres jet

### ISSUE THREAD

**Issue by K0stIa** [higerra/ridi_imu#7]
Hi, I am trying to build your cpp code, but getting the following errors:

`Building CXX object imu_localization/CMakeFiles/IMULocalization_cli.dir/imu_localization.cc.o
In file included from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:19:0,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/code/ridi_imu/cpp/imu_optimization/imu_optimization.h: In instantiation of ‘bool ridi::LocalSpeedFunctor<KVARIABLE, KCONSTRAINT>::operator()(const T*, const T*, const T*, T*) const [with T = ceres::Jet<double, 300>; int KVARIABLE = 100; int KCONSTRAINT = 200]’:
/home/k0stia/libraries/ceres/include/ceres/internal/variadic_evaluate.h:156:19:   required from ‘static bool ceres::internal::VariadicEvaluate<Functor, T, N0, N1, N2, 0, 0, 0, 0, 0, 0, 0>::Call(const Functor&, const T* const*, T*) [with Functor = ridi::LocalSpeedFunctor<100, 200>; T = ceres::Jet<double, 300>; int N0 = 100; int N1 = 100; int N2 = 100]’
/home/k0stia/libraries/ceres/include/ceres/internal/autodiff.h:289:72:   required from ‘static bool ceres::internal::AutoDiff<Functor, T, N0, N1, N2, N3, N4, N5, N6, N7, N8, N9>::Differentiate(const Functor&, const T* const*, int, T*, T**) [with Functor = ridi::LocalSpeedFunctor<100, 200>; T = double; int N0 = 100; int N1 = 100; int N2 = 100; int N3 = 0; int N4 = 0; int N5 = 0; int N6 = 0; int N7 = 0; int N8 = 0; int N9 = 0]’
/home/k0stia/libraries/ceres/include/ceres/autodiff_cost_function.h:211:66:   required from ‘bool ceres::AutoDiffCostFunction<CostFunctor, kNumResiduals, N0, N1, N2, N3, N4, N5, N6, N7, N8, N9>::Evaluate(const double* const*, double*, double**) const [with CostFunctor = ridi::LocalSpeedFunctor<100, 200>; int kNumResiduals = 600; int N0 = 100; int N1 = 100; int N2 = 100; int N3 = 0; int N4 = 0; int N5 = 0; int N6 = 0; int N7 = 0; int N8 = 0; int N9 = 0]’
/home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:267:1:   required from here
/home/k0stia/code/ridi_imu/cpp/imu_optimization/imu_optimization.h:122:46: error: no match for ‘operator*’ (operand types are ‘double’ and ‘Eigen::Matrix<ceres::Jet<double, 300>, 3, 1, 0, 3, 1>’)
           linacce_[i] + grid_->GetAlphaAt(i) * Eigen::Matrix<T, 3, 1>(bx[inv_ind], by[inv_ind], bz[inv_ind]);
                                              ^
In file included from /home/k0stia/libraries/eigen3.2.10/Eigen/Householder:17:0,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/QR:10,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Dense:4,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Eigen:1,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:14,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Householder/HouseholderSequence.h:410:99: note: candidate: template<class OtherDerived, class VectorsType, class CoeffsType, int Side> typename Eigen::internal::matrix_type_times_scalar_type<typename VectorsType::Scalar, OtherDerived>::Type Eigen::operator*(const Eigen::MatrixBase<Derived>&, const Eigen::HouseholderSequence<VectorsType, CoeffsType, Side>&)
 typename internal::matrix_type_times_scalar_type<typename VectorsType::Scalar,OtherDerived>::Type operator*(const MatrixBase<OtherDerived>& othe
                                                                                                   ^
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Householder/HouseholderSequence.h:410:99: note:   template argument deduction/substitution failed:
In file included from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:19:0,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/code/ridi_imu/cpp/imu_optimization/imu_optimization.h:122:46: note:   mismatched types ‘const Eigen::MatrixBase<Derived>’ and ‘double’
           linacce_[i] + grid_->GetAlphaAt(i) * Eigen::Matrix<T, 3, 1>(bx[inv_ind], by[inv_ind], bz[inv_ind]);
                                              ^
In file included from /home/k0stia/libraries/eigen3.2.10/Eigen/Core:322:0,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Dense:1,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Eigen:1,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:14,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/ProductBase.h:228:1: note: candidate: template<class Derived, class Lhs, class Rhs> typename Eigen::internal::enable_if<(! Eigen::internal::is_same<typename Derived::Scalar, typename Derived::RealScalar>::value), const Eigen::ScaledProduct<Derived> >::type Eigen::operator*(const typename Derived::RealScalar&, const Eigen::ProductBase<Derived, _Lhs, _Rhs>&)
 operator*(const typename Derived::RealScalar& x,const ProductBase<Derived,Lhs,Rhs>& prod)
 ^
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/ProductBase.h:228:1: note:   template argument deduction/substitution failed:
In file included from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:19:0,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/code/ridi_imu/cpp/imu_optimization/imu_optimization.h:122:46: note:   ‘Eigen::Matrix<ceres::Jet<double, 300>, 3, 1, 0, 3, 1>’ is not derived from ‘const Eigen::ProductBase<Derived, _Lhs, _Rhs>’
           linacce_[i] + grid_->GetAlphaAt(i) * Eigen::Matrix<T, 3, 1>(bx[inv_ind], by[inv_ind], bz[inv_ind]);
                                              ^
In file included from /home/k0stia/libraries/eigen3.2.10/Eigen/Core:322:0,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Dense:1,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Eigen:1,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:14,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/ProductBase.h:222:1: note: candidate: template<class Derived, class Lhs, class Rhs> const Eigen::ScaledProduct<Derived> Eigen::operator*(const typename Derived::Scalar&, const Eigen::ProductBase<Derived, _Lhs, _Rhs>&)
 operator*(const typename Derived::Scalar& x,const ProductBase<Derived,Lhs,Rhs>& prod)
 ^
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/ProductBase.h:222:1: note:   template argument deduction/substitution failed:
In file included from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:19:0,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/code/ridi_imu/cpp/imu_optimization/imu_optimization.h:122:46: note:   ‘Eigen::Matrix<ceres::Jet<double, 300>, 3, 1, 0, 3, 1>’ is not derived from ‘const Eigen::ProductBase<Derived, _Lhs, _Rhs>’
           linacce_[i] + grid_->GetAlphaAt(i) * Eigen::Matrix<T, 3, 1>(bx[inv_ind], by[inv_ind], bz[inv_ind]);
                                              ^
In file included from /home/k0stia/libraries/eigen3.2.10/Eigen/Core:322:0,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Dense:1,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Eigen:1,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:14,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/ProductBase.h:216:1: note: candidate: template<class Derived, class Lhs, class Rhs> typename Eigen::internal::enable_if<(! Eigen::internal::is_same<typename Derived::Scalar, typename Derived::RealScalar>::value), const Eigen::ScaledProduct<Derived> >::type Eigen::operator*(const Eigen::ProductBase<Derived, _Lhs, _Rhs>&, const typename Derived::RealScalar&)
 operator*(const ProductBase<Derived,Lhs,Rhs>& prod, const typename Derived::RealScalar& x)
 ^
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/ProductBase.h:216:1: note:   template argument deduction/substitution failed:
In file included from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:19:0,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/code/ridi_imu/cpp/imu_optimization/imu_optimization.h:122:46: note:   mismatched types ‘const Eigen::ProductBase<Derived, _Lhs, _Rhs>’ and ‘double’
           linacce_[i] + grid_->GetAlphaAt(i) * Eigen::Matrix<T, 3, 1>(bx[inv_ind], by[inv_ind], bz[inv_ind]);
                                              ^
In file included from /home/k0stia/libraries/eigen3.2.10/Eigen/Core:322:0,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Dense:1,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Eigen:1,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:14,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/ProductBase.h:210:1: note: candidate: template<class Derived, class Lhs, class Rhs> const Eigen::ScaledProduct<Derived> Eigen::operator*(const Eigen::ProductBase<Derived, _Lhs, _Rhs>&, const typename Derived::Scalar&)
 operator*(const ProductBase<Derived,Lhs,Rhs>& prod, const typename Derived::Scalar& x)
 ^
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/ProductBase.h:210:1: note:   template argument deduction/substitution failed:
In file included from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:19:0,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/code/ridi_imu/cpp/imu_optimization/imu_optimization.h:122:46: note:   mismatched types ‘const Eigen::ProductBase<Derived, _Lhs, _Rhs>’ and ‘double’
           linacce_[i] + grid_->GetAlphaAt(i) * Eigen::Matrix<T, 3, 1>(bx[inv_ind], by[inv_ind], bz[inv_ind]);
                                              ^
In file included from /home/k0stia/libraries/eigen3.2.10/Eigen/Core:314:0,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Dense:1,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Eigen:1,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:14,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/Transpositions.h:344:1: note: candidate: template<class Derived, class TranspositionDerived> const Eigen::internal::transposition_matrix_product_retval<TranspositionDerived, Derived, 1> Eigen::operator*(const Eigen::TranspositionsBase<TranspositionsDerived>&, const Eigen::MatrixBase<Derived>&)
 operator*(const TranspositionsBase<TranspositionDerived> &transpositions,
 ^
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/Transpositions.h:344:1: note:   template argument deduction/substitution failed:
In file included from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:19:0,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/code/ridi_imu/cpp/imu_optimization/imu_optimization.h:122:46: note:   mismatched types ‘const Eigen::TranspositionsBase<TranspositionsDerived>’ and ‘double’
           linacce_[i] + grid_->GetAlphaAt(i) * Eigen::Matrix<T, 3, 1>(bx[inv_ind], by[inv_ind], bz[inv_ind]);
                                              ^
In file included from /home/k0stia/libraries/eigen3.2.10/Eigen/Core:314:0,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Dense:1,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Eigen:1,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:14,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/Transpositions.h:331:1: note: candidate: template<class Derived, class TranspositionsDerived> const Eigen::internal::transposition_matrix_product_retval<TranspositionsDerived, Derived, 2> Eigen::operator*(const Eigen::MatrixBase<Derived>&, const Eigen::TranspositionsBase<TranspositionsDerived>&)
 operator*(const MatrixBase<Derived>& matrix,
 ^
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/Transpositions.h:331:1: note:   template argument deduction/substitution failed:
In file included from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:19:0,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/code/ridi_imu/cpp/imu_optimization/imu_optimization.h:122:46: note:   mismatched types ‘const Eigen::MatrixBase<Derived>’ and ‘double’
           linacce_[i] + grid_->GetAlphaAt(i) * Eigen::Matrix<T, 3, 1>(bx[inv_ind], by[inv_ind], bz[inv_ind]);
                                              ^
In file included from /home/k0stia/libraries/eigen3.2.10/Eigen/Core:313:0,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Dense:1,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Eigen:1,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:14,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/PermutationMatrix.h:552:1: note: candidate: template<class Derived, class PermutationDerived> const Eigen::internal::permut_matrix_product_retval<PermutationDerived, Derived, 1> Eigen::operator*(const Eigen::PermutationBase<PermutationDerived>&, const Eigen::MatrixBase<Derived>&)
 operator*(const PermutationBase<PermutationDerived> &permutation,
 ^
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/PermutationMatrix.h:552:1: note:   template argument deduction/substitution failed:
In file included from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:19:0,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/code/ridi_imu/cpp/imu_optimization/imu_optimization.h:122:46: note:   mismatched types ‘const Eigen::PermutationBase<PermutationDerived>’ and ‘double’
           linacce_[i] + grid_->GetAlphaAt(i) * Eigen::Matrix<T, 3, 1>(bx[inv_ind], by[inv_ind], bz[inv_ind]);
                                              ^
In file included from /home/k0stia/libraries/eigen3.2.10/Eigen/Core:313:0,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Dense:1,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Eigen:1,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:14,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/PermutationMatrix.h:539:1: note: candidate: template<class Derived, class PermutationDerived> const Eigen::internal::permut_matrix_product_retval<PermutationDerived, Derived, 2> Eigen::operator*(const Eigen::MatrixBase<Derived>&, const Eigen::PermutationBase<PermutationDerived>&)
 operator*(const MatrixBase<Derived>& matrix,
 ^
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/PermutationMatrix.h:539:1: note:   template argument deduction/substitution failed:
In file included from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:19:0,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/code/ridi_imu/cpp/imu_optimization/imu_optimization.h:122:46: note:   mismatched types ‘const Eigen::MatrixBase<Derived>’ and ‘double’
           linacce_[i] + grid_->GetAlphaAt(i) * Eigen::Matrix<T, 3, 1>(bx[inv_ind], by[inv_ind], bz[inv_ind]);
                                              ^
In file included from /home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/MatrixBase.h:136:0,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Core:279,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Dense:1,
                 from /home/k0stia/libraries/eigen3.2.10/Eigen/Eigen:1,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:14,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/../plugins/CommonCwiseUnaryOps.h:80:1: note: candidate: const Eigen::CwiseUnaryOp<Eigen::internal::scalar_multiple2_op<ceres::Jet<double, 300>, std::complex<ceres::Jet<double, 300> > >, const Eigen::Matrix<ceres::Jet<double, 300>, 3, 1, 0, 3, 1> > Eigen::operator*(const std::complex<ceres::Jet<double, 300> >&, const StorageBaseType&)
 operator*(const std::complex<Scalar>& scalar, const StorageBaseType& matrix)
 ^
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/../plugins/CommonCwiseUnaryOps.h:80:1: note:   no known conversion for argument 1 from ‘double’ to ‘const std::complex<ceres::Jet<double, 300> >&’
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/../plugins/CommonCwiseUnaryOps.h:76:1: note: candidate: const ScalarMultipleReturnType Eigen::operator*(const Scalar&, const StorageBaseType&)
 operator*(const Scalar& scalar, const StorageBaseType& matrix)
 ^
/home/k0stia/libraries/eigen3.2.10/Eigen/src/Core/../plugins/CommonCwiseUnaryOps.h:76:1: note:   no known conversion for argument 1 from ‘double’ to ‘const Scalar& {aka const ceres::Jet<double, 300>&}’
In file included from /home/k0stia/libraries/ceres/include/ceres/internal/autodiff.h:145:0,
                 from /home/k0stia/libraries/ceres/include/ceres/autodiff_cost_function.h:132,
                 from /home/k0stia/libraries/ceres/include/ceres/ceres.h:37,
                 from /home/k0stia/code/ridi_imu/cpp/imu_optimization/imu_optimization.h:12,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:19,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/libraries/ceres/include/ceres/jet.h:368:11: note: candidate: template<class T, int N> ceres::Jet<T, N> ceres::operator*(T, const ceres::Jet<T, N>&)
 Jet<T, N> operator*(T s, const Jet<T, N>& f) {
           ^
/home/k0stia/libraries/ceres/include/ceres/jet.h:368:11: note:   template argument deduction/substitution failed:
In file included from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:19:0,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/code/ridi_imu/cpp/imu_optimization/imu_optimization.h:122:46: note:   ‘Eigen::Matrix<ceres::Jet<double, 300>, 3, 1, 0, 3, 1>’ is not derived from ‘const ceres::Jet<T, N>’
           linacce_[i] + grid_->GetAlphaAt(i) * Eigen::Matrix<T, 3, 1>(bx[inv_ind], by[inv_ind], bz[inv_ind]);
                                              ^
In file included from /home/k0stia/libraries/ceres/include/ceres/internal/autodiff.h:145:0,
                 from /home/k0stia/libraries/ceres/include/ceres/autodiff_cost_function.h:132,
                 from /home/k0stia/libraries/ceres/include/ceres/ceres.h:37,
                 from /home/k0stia/code/ridi_imu/cpp/imu_optimization/imu_optimization.h:12,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:19,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/libraries/ceres/include/ceres/jet.h:362:11: note: candidate: template<class T, int N> ceres::Jet<T, N> ceres::operator*(const ceres::Jet<T, N>&, T)
 Jet<T, N> operator*(const Jet<T, N>& f, T s) {
           ^
/home/k0stia/libraries/ceres/include/ceres/jet.h:362:11: note:   template argument deduction/substitution failed:
In file included from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:19:0,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/code/ridi_imu/cpp/imu_optimization/imu_optimization.h:122:46: note:   mismatched types ‘const ceres::Jet<T, N>’ and ‘double’
           linacce_[i] + grid_->GetAlphaAt(i) * Eigen::Matrix<T, 3, 1>(bx[inv_ind], by[inv_ind], bz[inv_ind]);
                                              ^
In file included from /home/k0stia/libraries/ceres/include/ceres/internal/autodiff.h:145:0,
                 from /home/k0stia/libraries/ceres/include/ceres/autodiff_cost_function.h:132,
                 from /home/k0stia/libraries/ceres/include/ceres/ceres.h:37,
                 from /home/k0stia/code/ridi_imu/cpp/imu_optimization/imu_optimization.h:12,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:19,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/libraries/ceres/include/ceres/jet.h:355:11: note: candidate: template<class T, int N> ceres::Jet<T, N> ceres::operator*(const ceres::Jet<T, N>&, const ceres::Jet<T, N>&)
 Jet<T, N> operator*(const Jet<T, N>& f,
           ^
/home/k0stia/libraries/ceres/include/ceres/jet.h:355:11: note:   template argument deduction/substitution failed:
In file included from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.h:19:0,
                 from /home/k0stia/code/ridi_imu/cpp/imu_localization/imu_localization.cc:5:
/home/k0stia/code/ridi_imu/cpp/imu_optimization/imu_optimization.h:122:46: note:   mismatched types ‘const ceres::Jet<T, N>’ and ‘double’
           linacce_[i] + grid_->GetAlphaAt(i) * Eigen::Matrix<T, 3, 1>(bx[inv_ind], by[inv_ind], bz[inv_ind]);
                                              ^

                                                        ^
imu_localization/CMakeFiles/IMULocalization_cli.dir/build.make:62: recipe for target 'imu_localization/CMakeFiles/IMULocalization_cli.dir/imu_localization.cc.o' failed
make[2]: *** [imu_localization/CMakeFiles/IMULocalization_cli.dir/imu_localization.cc.o] Error 1
CMakeFiles/Makefile2:365: recipe for target 'imu_localization/CMakeFiles/IMULocalization_cli.dir/all' failed
make[1]: *** [imu_localization/CMakeFiles/IMULocalization_cli.dir/all] Error 2
Makefile:94: recipe for target 'all' failed
make: *** [all] Error 2
`

**Comment by K0stIa** [comment]
I have tried gcc and clang, both give similar error related to Ceres::Jet.

**Comment by higerra** [comment]
What OS do you use and how did you install Ceres solver? Looks like the library is not installed in standard path (e.g. /usr/local), so the linker might not be able to find the library file.

**Comment by fangxu622** [comment]
ubuntu 16.04 , ceres is installed in standard path ：/usr/local/ ,but I get same problem ,  the ceres version 3.2.92

**Comment by fangxu622** [comment]
and i have a question when run "cmake .." the output  indicate glog not found .like this. Does it affect this matter?
```
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found OpenCV: /home/fangxu/Documents/opencv34build (found suitable version "3.4.0", minimum required is "3.0") 
-- Found Eigen: /usr/include/eigen3 (found version "3.2.92") 
-- No preference for use of exported glog CMake configuration set, and no hints for include/library directories provided. Defaulting to preferring an installed/exported glog CMake configuration if available.
-- Failed to find installed glog CMake configuration, searching for glog build directories exported with CMake.
-- Failed to find an installed/exported CMake configuration for glog, will perform search for installed glog components.
-- Found Glog: /usr/local/include  
-- No preference for use of exported gflags CMake configuration set, and no hints for include/library directories provided. Defaulting to preferring an installed/exported gflags CMake configuration if available.
-- Found installed version of gflags: /usr/local/lib/cmake/gflags
-- Detected gflags version: 2.2.2
-- Found Gflags: /usr/local/include  
-- Opencv lib status:
-- version: 3.4.0
-- lib: opencv_calib3d;opencv_core;opencv_dnn;opencv_features2d;opencv_flann;opencv_highgui;opencv_imgcodecs;opencv_imgproc;opencv_ml;opencv_objdetect;opencv_photo;opencv_shape;opencv_stitching;opencv_superres;opencv_video;opencv_videoio;opencv_videostab;opencv_viz;opencv_aruco;opencv_bgsegm;opencv_bioinspired;opencv_ccalib;opencv_datasets;opencv_dpm;opencv_face;opencv_freetype;opencv_fuzzy;opencv_hdf;opencv_img_hash;opencv_line_descriptor;opencv_optflow;opencv_phase_unwrapping;opencv_plot;opencv_reg;opencv_rgbd;opencv_saliency;opencv_sfm;opencv_stereo;opencv_structured_light;opencv_surface_matching;opencv_text;opencv_tracking;opencv_xfeatures2d;opencv_ximgproc;opencv_xobjdetect;opencv_xphoto
-- include path: /home/fangxu/Documents/opencv34build/include;/home/fangxu/Documents/opencv34build/include/opencv
-- Glog lib status:
-- Glog version: 
-- Glog lib: 
-- include path: 
-- Found OpenMP
-- Failed to find installed glog CMake configuration, searching for glog build directories exported with CMake.
-- Failed to find an installed/exported CMake configuration for glog, will perform search for installed glog components.
-- Failed to find installed glog CMake configuration, searching for glog build directories exported with CMake.
-- Failed to find an installed/exported CMake configuration for glog, will perform search for installed glog components.
-- Found installed version of gflags: /usr/local/lib/cmake/gflags
-- Detected gflags version: 2.2.2
-- Found OpenCV: /home/fangxu/Documents/opencv34build (found version "3.4.0") 
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found installed version of Eigen: /usr/lib/cmake/eigen3
-- Found required Ceres dependency: Eigen version 3.2.92 in /usr/include/eigen3
-- Failed to find installed glog CMake configuration, searching for glog build directories exported with CMake.
-- Failed to find an installed/exported CMake configuration for glog, will perform search for installed glog components.
-- Found required Ceres dependency: glog
-- Found installed version of gflags: /usr/local/lib/cmake/gflags
-- Detected gflags version: 2.2.2
-- Found required Ceres dependency: gflags
-- Found Ceres version: 2.0.0 installed in: /usr/local with components: [EigenSparse, SparseLinearAlgebraLibrary, LAPACK, SuiteSparse, CXSparse, SchurSpecializations]
-- Failed to find installed glog CMake configuration, searching for glog build directories exported with CMake.
-- Failed to find an installed/exported CMake configuration for glog, will perform search for installed glog components.
-- Found installed version of gflags: /usr/local/lib/cmake/gflags
-- Detected gflags version: 2.2.2
-- Failed to find installed glog CMake configuration, searching for glog build directories exported with CMake.
-- Failed to find an installed/exported CMake configuration for glog, will perform search for installed glog components.
-- Found installed version of gflags: /usr/local/lib/cmake/gflags
-- Detected gflags version: 2.2.2
-- Failed to find installed glog CMake configuration, searching for glog build directories exported with CMake.
-- Failed to find an installed/exported CMake configuration for glog, will perform search for installed glog components.
-- Found installed version of gflags: /usr/local/lib/cmake/gflags
-- Detected gflags version: 2.2.2
-- Found installed version of Eigen: /usr/lib/cmake/eigen3
-- Found required Ceres dependency: Eigen version 3.2.92 in /usr/include/eigen3
-- Failed to find installed glog CMake configuration, searching for glog build directories exported with CMake.
-- Failed to find an installed/exported CMake configuration for glog, will perform search for installed glog components.
-- Found required Ceres dependency: glog
-- Found installed version of gflags: /usr/local/lib/cmake/gflags
-- Detected gflags version: 2.2.2
-- Found required Ceres dependency: gflags
-- Found Ceres version: 2.0.0 installed in: /usr/local with components: [EigenSparse, SparseLinearAlgebraLibrary, LAPACK, SuiteSparse, CXSparse, SchurSpecializations]
-- Configuring done
-- Generating done

```
![image](https://user-images.githubusercontent.com/16681534/53474168-59ad7480-3aa7-11e9-8448-b945ece1bc8c.png)

**Comment by ap-antoniuk-kostia** [comment]
Sorry, I did not reply here. I have same problems when ceres is installed in standard way. 
I gave up on compiling this code. If you guys solve this problem, let me know please.

cheers, -K.

**Comment by K0stIa** [comment]
Sorry, I did not reply here. I have same problems when ceres is installed in standard way.
I gave up on compiling this code. If you guys solve this problem, let me know please.

cheers, -K.

**Comment by fangxu622** [comment]
@higerra  能说一下你相应运行的库的版本的情况吗

**Comment by fangxu622** [comment]
@K0stIa  I fix it ,you shoud install eigen version  3.3  above ，then compile ceres-solvers, it took me a whole day!

**Comment by higerra** [comment]
Thank you very much for your verification! I will clarify Eigen requirement.
